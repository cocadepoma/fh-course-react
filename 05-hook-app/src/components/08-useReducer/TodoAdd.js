import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        handleAddTodo({
            id: new Date().getTime(),
            desc: description,
            done: false,
        });

        // reset();
    };
    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit} className='d-flex flex-column'>
                <input
                    onChange={handleInputChange}
                    className='form-control'
                    type='text'
                    name='description'
                    value={description}
                    placeholder='Aprender...'
                    autoComplete='off'
                />

                <button type='submit' className='btn btn-block btn-outline-primary mt-1'>
                    Agregar
                </button>
            </form>
        </>
    );
};
