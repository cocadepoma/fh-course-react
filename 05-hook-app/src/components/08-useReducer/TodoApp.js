import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import "./styles.css";

export const TodoApp = () => {
    // TODOAPP LIST

    const init = () => {
        return JSON.parse(localStorage.getItem("todos")) || [];
        // return [
        //     { id: new Date().getTime(), desc: "Aprender React", done: false },
        // ];
    };

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (id) => {
        const action = {
            type: "delete",
            payload: id,
        };
        dispatch(action);
    };

    const handleToggle = (id) => {
        dispatch({
            type: "toggle",
            payload: id,
        });
    };

    const handleAddTodo = (desc) => {
        dispatch({
            type: "add",
            payload: desc,
        });
    };
    return (
        <div>
            <h1>Todo APP ({todos.length})</h1>
            <hr />
            <div className='row'>
                <div className='col-md-7'>
                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                </div>
                <div className='col-md-5'>
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    );
};
