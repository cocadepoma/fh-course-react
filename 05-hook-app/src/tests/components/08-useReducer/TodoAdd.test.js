import { shallow } from "enzyme";
import React from "react";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe("pruebas en <TodoApp/>", () => {
    const handleAddTodo = jest.fn();

    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

    test("debe de mostrarse correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("No debe de llamar handleAddTodo", () => {
        //wrapper.find("button").simulate("click");
        const formSubmit = wrapper.find("form").prop("onSubmit");
        formSubmit({
            preventDefault() {},
        });

        expect(handleAddTodo).not.toHaveBeenCalled();
        // Las 2 podrían servir
        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test("debe de llamar la función handleAddTodo", () => {
        const value = "Aprender Ionic";

        wrapper.find("input").simulate("change", {
            target: {
                value,
                name: "description",
            },
        });

        const formSubmit = wrapper.find("form").prop("onSubmit");
        formSubmit({
            preventDefault() {},
        });

        expect(handleAddTodo).toHaveBeenCalled();
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false,
        });

        expect(wrapper.find("input").prop("value")).toBe(value);
    });
});
