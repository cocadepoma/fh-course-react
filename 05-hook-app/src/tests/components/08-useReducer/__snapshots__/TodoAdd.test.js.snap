// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoApp/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    className="d-flex flex-column"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-block btn-outline-primary mt-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
